### PR TITLE
Add methods and tests for TreeEntry

### PR DESCRIPTION
**Pull request by kshyatt**

### CODE REVIEW

File: base/libgit2/tree.jl [hunk 1]
```diff
@@ -34,12 +34,28 @@ end
 
 function entrytype(te::GitTreeEntry)
     otype = ccall((:git_tree_entry_type, :libgit2), Cint, (Ptr{Void},), te.ptr)
-    return getobjecttype(otype)
+    return objtype(Consts.OBJECT(otype))
 end
 
 function entryid(te::GitTreeEntry)
-    oid_ptr = ccall((:git_tree_entry_id, :libgit2), Cint, (Ptr{Void},), te.ptr)
-    return GitHash(oid_ptr[])
```
Comment (tkelman) [inline base/libgit2/tree.jl]: did this previous code work at all?

File: base/libgit2/tree.jl [hunk 2]
```diff
@@ -34,12 +34,28 @@ end
...
+    oid_ptr = ccall((:git_tree_entry_id, :libgit2), Ptr{UInt8}, (Ptr{Void},), te.ptr)
+    oid_ptr == C_NULL && return GitHash()
```
Comment (tkelman) [inline base/libgit2/tree.jl]: I thought we were moving away from returning empty githashes?

File: base/libgit2/tree.jl [hunk 3]
```diff
@@ -34,12 +34,28 @@ end
...
+    return GitHash(oid_ptr)
+end
+
+function Base.count(tree::GitTree)
+    return ccall((:git_tree_entrycount, :libgit2), Csize_t, (Ptr{Void},), tree.ptr)
+end
+
+function Base.getindex(tree::GitTree, i::Integer)
+    if i < 1 || i > count(tree)
+        throw(BoundsError())
```
Comment (tkelman) [inline base/libgit2/tree.jl]: include object and index?

File: base/libgit2/tree.jl [hunk 4]
```diff
@@ -34,12 +34,28 @@ end
...
+    end
+    te_ptr = ccall((:git_tree_entry_byindex, :libgit2),
+                   Ptr{Void},
+                   (Ptr{Void}, Csize_t), tree.ptr, i-1)
+    te_ptr == C_NULL && return nothing
```
Comment (tkelman) [inline base/libgit2/tree.jl]: also thought we were trying to get away from silently returning nothing

File: base/libgit2/tree.jl (hunk 1 above)
Comment (kshyatt) [inline base/libgit2/tree.jl]: i don't think so

**Comment by kshyatt** [comment]
Somewhat better?

**Review by ararslan — APPROVED** [review]
LGTM

**Comment by kshyatt** [comment]
@tkelman is this ok with you?

**Comment by tkelman** [comment]
looks okay, but probably worth checking whether PkgDev relies on this in any not-tested-by-base ways

**Comment by kshyatt** [comment]
I checked `PkgDev` just now for any calls to `entryid`/`entrytype`/any `TreeEntry` functions and nothing.

**Pull request merged**